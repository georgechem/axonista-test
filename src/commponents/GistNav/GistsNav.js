import React from 'react';
import PropTypes from 'prop-types';

import svgSortOrder from './icons/sort_black_24dp.svg';
import svgByDate from './icons/date_range_black_24dp.svg';
import svgByDescription from './icons/description_black_24dp.svg';
import svgByPerson from './icons/person_black_24dp.svg';

import './gistNav.scss';

const GistsNav = (props) => {

    const onSortingOrderChange = () => {
        if(props.sortingOrder.direction === 'desc') props.setSortingOrder({direction: 'asc'});
        else props.setSortingOrder({direction: 'desc'});
        const el = document.getElementById('sortOrderBtn');
        el.classList.toggle('rotate180deg');
    }

    return (
        <div className="gistNav">
            <div className="gistNav__top">
                <div className="gistNav__top__search">
                    <input
                        data-testid="nav-input"
                        id="inputSearch"
                        className="gistNav__top__input"
                        type="text"
                        name="input"
                        title="user input"
                        onChange={(event) => props.onInputChange(event)}
                        onKeyDown={(event) => props.onBackSpaceClick(event)}
                    />
                    <select
                        data-testid="nav-select"
                        className="gistNav__top__select"
                        id="itemsPerPage"
                        name="itemsPerPage"
                        title="Items per page"
                        defaultValue={props.perPage}
                        onChange={(e)=>{props.onSelectPerPage(e)}}
                    >
                        <option data-testid="navOption5" value="5">5</option>
                        <option data-testid="navOption10" value="10">10</option>
                        <option data-testid="navOption30" value="30">30</option>
                        <option data-testid="navOption100" value="100">100</option>
                    </select>
                </div>
                <div className="gistNav__top__buttons">
                    <img
                        data-testid="nav-sortOrderBtn"
                        id="sortOrderBtn"
                        title="Sort Order"
                        className="gistNav__top__buttons__sortOrder rotate180deg"
                        src={svgSortOrder}
                        onClick={onSortingOrderChange}
                    alt="icon order"/>
                    <div className="gistNav__top__buttons__switchGroup">
                        <p><small data-testid="nav-sortingAttribute">{props.sortingAttribute?.sortingCategory}</small></p>
                        <img
                            data-testid="nav-sortByDateSwitch"
                            id="sortByDateSwitch"
                            alt="icon of calendar"
                            title="Sort By Date"
                            className="gistNav__top__buttons__switchGroup__switch"
                            src={svgByDate}
                            onClick={(e) => props.onSortingAttributeChange(e, {sortingCategory: 'date'})}
                        />
                        <img
                            data-testid="nav-sortByDescriptionSwitch"
                            id="sortByDescriptionSwitch"
                            alt="icon of paper sheet"
                            title="Sort By Description"
                            className="gistNav__top__buttons__switchGroup__switch"
                            src={svgByDescription}
                            onClick={(e) => props.onSortingAttributeChange(e, {sortingCategory: 'description'})}
                        />
                        <img
                            data-testid="nav-sortByLoginSwitch"
                            id="sortByLoginSwitch"
                            alt="icon of user"
                            title="Sort By Login"
                            className="gistNav__top__buttons__switchGroup__switch"
                            src={svgByPerson}
                            onClick={(e) => props.onSortingAttributeChange(e, {sortingCategory: 'login'})}
                        />
                    </div>
                </div>
            </div>
        </div>
    );
};

GistsNav.propTypes = {
    perPage: PropTypes.number.isRequired,
    onSortingAttributeChange: PropTypes.func.isRequired,
    onInputChange: PropTypes.func.isRequired,
    onBackSpaceClick: PropTypes.func.isRequired,
    onSelectPerPage: PropTypes.func.isRequired,
    sortingOrder: PropTypes.object.isRequired,
    setSortingOrder: PropTypes.func.isRequired
}

export default GistsNav;