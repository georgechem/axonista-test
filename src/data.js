export const data = {
    "url": "https://api.github.com/gists/61214c3cd9da6076fc6f443ba310d10a",
    "forks_url": "https://api.github.com/gists/61214c3cd9da6076fc6f443ba310d10a/forks",
    "commits_url": "https://api.github.com/gists/61214c3cd9da6076fc6f443ba310d10a/commits",
    "id": "61214c3cd9da6076fc6f443ba310d10a",
    "node_id": "G_kwDOADF-UdoAIDYxMjE0YzNjZDlkYTYwNzZmYzZmNDQzYmEzMTBkMTBh",
    "git_pull_url": "https://gist.github.com/61214c3cd9da6076fc6f443ba310d10a.git",
    "git_push_url": "https://gist.github.com/61214c3cd9da6076fc6f443ba310d10a.git",
    "html_url": "https://gist.github.com/61214c3cd9da6076fc6f443ba310d10a",
    "files": {
        "as": {
            "filename": "as",
            "type": "text/plain",
            "language": null,
            "raw_url": "https://gist.githubusercontent.com/annakhloyan/61214c3cd9da6076fc6f443ba310d10a/raw/2f259b79aa7e263f5829bb6e98096e7ec976d998/as",
            "size": 1
        }
    },
    "public": true,
    "created_at": "2021-09-25T17:26:24Z",
    "updated_at": "2021-09-25T17:26:24Z",
    "description": "1w",
    "comments": 0,
    "user": null,
    "comments_url": "https://api.github.com/gists/61214c3cd9da6076fc6f443ba310d10a/comments",
    "owner": {
        "login": "annakhloyan",
        "id": 3243601,
        "node_id": "MDQ6VXNlcjMyNDM2MDE=",
        "avatar_url": "https://avatars.githubusercontent.com/u/3243601?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/annakhloyan",
        "html_url": "https://github.com/annakhloyan",
        "followers_url": "https://api.github.com/users/annakhloyan/followers",
        "following_url": "https://api.github.com/users/annakhloyan/following{/other_user}",
        "gists_url": "https://api.github.com/users/annakhloyan/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/annakhloyan/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/annakhloyan/subscriptions",
        "organizations_url": "https://api.github.com/users/annakhloyan/orgs",
        "repos_url": "https://api.github.com/users/annakhloyan/repos",
        "events_url": "https://api.github.com/users/annakhloyan/events{/privacy}",
        "received_events_url": "https://api.github.com/users/annakhloyan/received_events",
        "type": "User",
        "site_admin": false
    },
    "truncated": false
};