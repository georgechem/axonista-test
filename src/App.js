import React, {useState, useEffect} from 'react';
import GistsNav from "./commponents/GistNav/GistsNav";
import GistsContent from "./commponents/GistContent/GistsContent";
import GistPagination from "./commponents/GistPagination/GistPagination";
import './style.scss';


const App = () => {
    const [ready, setReady] = useState(false);
    const [gists, setGists] = useState([]);
    const [gistsCopy, setGistsCopy] = useState([]);
    const [page, setPage] = useState(1);
    const [perPage, setPerPage] = useState(10);
    const [content, setContent] = useState('');
    const [sortingOrder, setSortingOrder] = useState({direction: 'desc'});
    const [sortingAttribute, setSortingAttribute] = useState({sortingCategory: 'login'});


    /**
     * Fetch data from Github API
     */
    useEffect(()=>{
        const url = `https://api.github.com/gists?page=${page}&per_page=${perPage}`;
        fetch(url,{
            method: 'GET',
            cache: 'no-cache',
            headers: {
                'Content-Type': 'application/json'
            }
        })
            .then(res => res.json())
            .then(data => {
                // Sort data after fetching
                const dataSorted = sortGists(true, data);
                setGists(dataSorted);
                setGistsCopy(dataSorted);
                // Set state that data are fetched and ready
                setReady(true);
            })
            .catch(err => {
                console.log(err);
            });
    },[page, perPage]);

    /**
     * Sort Gists as sorting order OR/AND sorting attribute change
     */
    useEffect(() => {
        //sortGists();
    }, [sortingOrder, sortingAttribute])

    /**
     * Match Gists to pattern as any changes to input content
     */
    useEffect(() => {
        //matchGists();
    },[content])

    /**
     * Rotate sorting order icon as sorting order change
     */
    useEffect(() => {
        const el = document.getElementById('sortOrderBtn');
        el.classList.toggle('rotate180deg');
    }, [sortingOrder]);

    /**
     * Make sure one of button is pressed when application starts
     */
    useEffect(()=>{
        // Fire first soring indirectly by switching state
        console.log('useEffect - onSortBy', gists);
        onSortBy('sortByLoginSwitch', sortingAttribute);
    },[]);

    useEffect(() => {
        //sortGists();
        console.log(`sorting gists with page`, gists)
    },[]);

    useEffect(()=>{

        // sort by user input
        matchGists();
        // sort by chosen icon
        sortGists();
        //setGists(gistsCopy);
        console.log('gist copy READY', gists);
    }, [ready, sortingOrder, sortingAttribute]);


    /**
     * array of elements ID for all SortBy Switches in Nav
     * @type {string[]}
     */
    const allSwitches = ['sortByDateSwitch', 'sortByDescriptionSwitch', 'sortByLoginSwitch'];

    /**
     * Remove certain class from all SortBy buttons
     */
    const turnOfAllSwitches = () => {
        allSwitches.forEach((switchButton) => {
            const el = document?.getElementById(switchButton);
            if(el !== null){
                el.classList.remove('turnedOn');
            }
        })
    }

    /**
     * Restore all gists as user modifying input value
     * @param event
     */
    const onBackSpaceClick = (event) => {
        if(event.key === 'Backspace' || event.code === 'Backspace'){
            setGists(gistsCopy);
        }
    }

    /**
     * Handle every change of content in input element | text sortingByDate excluded
     * @param event
     */
    const onInputChange = (event) => {
        // if icon date selected do not sort with input text
        if(sortingAttribute.sortingCategory === 'date') return;
        setContent(event?.target?.value || '');
    }

    /**
     * Set SortingAttribute & make button pressed by adding class
     * @param elementId
     * @param state
     */
    const onSortBy = (elementId, state) => {
        const el = document?.getElementById(elementId);
        if(el !== null){
            turnOfAllSwitches();
            el.classList.add('turnedOn');
            setSortingAttribute(state);
        }
    }

    /**
     * Filter Gits to match user pattern
     * @param data
     */
    const matchGists = (data) => {
        const tmp = [...gistsCopy];
        let findings = [];
        if(sortingAttribute.sortingCategory === 'login') findings = tmp.filter(gist => (gist?.owner?.login?.toLowerCase().search(content) !== -1));
        else if(sortingAttribute.sortingCategory === 'description') findings = tmp.filter(gist => gist?.description?.toLowerCase().search(content) !== -1);
        else findings = tmp.filter(gist => true);
        setGists(findings);
    };

    /**
     * Sorting gists by - LOGIN | DATE | DESCRIPTION and ASC | DESC
     */
    const sortGists = (fetchMode = false, data = []) => {
        let tmp =[];
        (!fetchMode) ? tmp = [...gistsCopy] : tmp = [...data];

        const compareStrings = (a, b) => {
            if(a < b) return -1;
            if(a > b) return 1;
            return 0;
        }

        if(sortingOrder.direction === 'desc'){
            if(sortingAttribute.sortingCategory === 'login') tmp.sort((a, b) => compareStrings(a?.owner?.login?.toLowerCase(), b?.owner?.login?.toLowerCase()));
            if(sortingAttribute.sortingCategory === 'date') tmp.sort((a, b) => compareStrings(a?.updated_at, b?.updated_at));
            if(sortingAttribute.sortingCategory === 'description') tmp.sort((a, b) => compareStrings(a?.description?.toLowerCase(), b?.description?.toLowerCase()));
        }else{
            if(sortingAttribute.sortingCategory === 'login') tmp.sort((a, b) => compareStrings(b?.owner?.login?.toLowerCase(), a?.owner?.login?.toLowerCase()));
            if(sortingAttribute.sortingCategory === 'date') tmp.sort((a, b) => compareStrings(b?.updated_at, a?.updated_at));
            if(sortingAttribute.sortingCategory === 'description') tmp.sort((a, b) => compareStrings(b?.description?.toLowerCase(), a?.description?.toLowerCase()));
        }
        console.log(`gists sorted in sortGist `, tmp);
        setGistsCopy(tmp);
        if(fetchMode) return tmp;
    }


    /**
     * NEXT page
     */
    const onNextPageClick = () => {
        if((page * perPage) >= 2999) return;
        setPage(page + 1);
    }
    /**
     * PREV page
     */
    const onPrevPageClick = () => {
        if(page === 1) return;
        setPage(page - 1);
    }
    /**
     * Set elements per Page
     */
    const onSelectPerPage = (e) => {
        const options = e.target.options;
        const index = e.target.options.selectedIndex;
        const value = parseInt(options[index].value) | 1;
        setPerPage(value);
    }



    return (
        <div>
            <GistsNav
                perPage={perPage}
                onSortBy={onSortBy}
                onInputChange={onInputChange}
                onBackSpaceClick={onBackSpaceClick}
                onSelectPerPageFn={onSelectPerPage}
                sortingOrder={sortingOrder} setSortingOrder={setSortingOrder}
            />
            {gists?.map((gist, key) => <GistsContent
                key={'gistKey_' + key}
                gist={gist}
            />)}
            <GistPagination
                onNextPageClick={onNextPageClick}
                page={page}
                onPrevPageClick={onPrevPageClick}
            />
        </div>
    );
};

export default App;