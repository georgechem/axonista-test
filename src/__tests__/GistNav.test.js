import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {cleanup, render, screen, fireEvent} from '@testing-library/react';
import '@testing-library/jest-dom';
import App from '../App';
import GistsNav from "../commponents/GistNav/GistsNav";
import {data} from "../data";

global.fetch = jest.fn(() =>
    Promise.resolve({
        json: ()=> Promise.resolve(data)
    })
);

afterEach(cleanup);

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<GistsNav
        perPage={5}
        onSortingAttributeChange={jest.fn()}
        onInputChange={jest.fn()}
        onBackSpaceClick={jest.fn()}
        onSelectPerPage={jest.fn()}
        sortingOrder={{}}
        setSortingOrder={jest.fn()}
    />, div);
});

test('matches Snapshot', () => {

    const tree = renderer.create(<GistsNav
        perPage={5}
        onSortingAttributeChange={jest.fn()}
        onInputChange={jest.fn()}
        onBackSpaceClick={jest.fn()}
        onSelectPerPage={jest.fn()}
        sortingOrder={{}}
        setSortingOrder={jest.fn()}
    />).toJSON();

    expect(tree).toMatchSnapshot();

});

test('initial sorting attribute is login and sorting attribute is working', ()=>{
    const {getByTestId} = render(
        <App>
            <GistsNav
                onBackSpaceClick={()=>jest.fn()}
                onInputChange={()=>jest.fn()}
                onSelectPerPage={()=>jest.fn()}
                onSortingAttributeChange={()=>jest.fn()}
                perPage={10}
                setSortingOrder={()=>jest.fn()}
                sortingOrder={{direction: 'desc'}}
                sortingAttribute={{}}
            />
        </App>
    );

    const sortingAttribute = getByTestId('nav-sortingAttribute');

    const navSortByDateSwitch = getByTestId('nav-sortByDateSwitch');
    const navSortByDescriptionSwitch = getByTestId('nav-sortByDescriptionSwitch');
    const navSortByLoginSwitch = getByTestId('nav-sortByLoginSwitch');

    expect(sortingAttribute.textContent).toBe('login');
    fireEvent.click(navSortByDateSwitch);
    expect(sortingAttribute.textContent).toBe('date');
    fireEvent.click(navSortByDescriptionSwitch);
    expect(sortingAttribute.textContent).toBe('description');
    fireEvent.click(navSortByLoginSwitch);
    expect(sortingAttribute.textContent).toBe('login');

});

test('select per page is 10', ()=>{
    const {getByTestId} = render(
        <App>
            <GistsNav
                onBackSpaceClick={()=>jest.fn()}
                onInputChange={()=>jest.fn()}
                onSelectPerPage={()=>jest.fn()}
                onSortingAttributeChange={()=>jest.fn()}
                perPage={10}
                setSortingOrder={()=>jest.fn()}
                sortingOrder={{direction: 'desc'}}
                sortingAttribute={{}}
            />
        </App>
    );

    const navOption10 = getByTestId('navOption10');

    expect(navOption10).toHaveAttribute('selected');
    expect(navOption10.textContent).toBe('10');

});