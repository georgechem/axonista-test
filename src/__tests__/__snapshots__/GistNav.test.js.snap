// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches Snapshot 1`] = `
<div
  className="gistNav"
>
  <div
    className="gistNav__top"
  >
    <div
      className="gistNav__top__search"
    >
      <input
        className="gistNav__top__input"
        data-testid="nav-input"
        id="inputSearch"
        name="input"
        onChange={[Function]}
        onKeyDown={[Function]}
        title="user input"
        type="text"
      />
      <select
        className="gistNav__top__select"
        data-testid="nav-select"
        defaultValue={5}
        id="itemsPerPage"
        name="itemsPerPage"
        onChange={[Function]}
        title="Items per page"
      >
        <option
          data-testid="navOption5"
          value="5"
        >
          5
        </option>
        <option
          data-testid="navOption10"
          value="10"
        >
          10
        </option>
        <option
          data-testid="navOption30"
          value="30"
        >
          30
        </option>
        <option
          data-testid="navOption100"
          value="100"
        >
          100
        </option>
      </select>
    </div>
    <div
      className="gistNav__top__buttons"
    >
      <img
        alt="icon order"
        className="gistNav__top__buttons__sortOrder rotate180deg"
        data-testid="nav-sortOrderBtn"
        id="sortOrderBtn"
        onClick={[Function]}
        src={
          Object {
            "canInstrument": true,
            "createTransformer": [Function],
            "getCacheKey": [Function],
            "getCacheKeyAsync": [Function],
            "process": [Function],
            "processAsync": [Function],
          }
        }
        title="Sort Order"
      />
      <div
        className="gistNav__top__buttons__switchGroup"
      >
        <p>
          <small
            data-testid="nav-sortingAttribute"
          />
        </p>
        <img
          alt="icon of calendar"
          className="gistNav__top__buttons__switchGroup__switch"
          data-testid="nav-sortByDateSwitch"
          id="sortByDateSwitch"
          onClick={[Function]}
          src={
            Object {
              "canInstrument": true,
              "createTransformer": [Function],
              "getCacheKey": [Function],
              "getCacheKeyAsync": [Function],
              "process": [Function],
              "processAsync": [Function],
            }
          }
          title="Sort By Date"
        />
        <img
          alt="icon of paper sheet"
          className="gistNav__top__buttons__switchGroup__switch"
          data-testid="nav-sortByDescriptionSwitch"
          id="sortByDescriptionSwitch"
          onClick={[Function]}
          src={
            Object {
              "canInstrument": true,
              "createTransformer": [Function],
              "getCacheKey": [Function],
              "getCacheKeyAsync": [Function],
              "process": [Function],
              "processAsync": [Function],
            }
          }
          title="Sort By Description"
        />
        <img
          alt="icon of user"
          className="gistNav__top__buttons__switchGroup__switch"
          data-testid="nav-sortByLoginSwitch"
          id="sortByLoginSwitch"
          onClick={[Function]}
          src={
            Object {
              "canInstrument": true,
              "createTransformer": [Function],
              "getCacheKey": [Function],
              "getCacheKeyAsync": [Function],
              "process": [Function],
              "processAsync": [Function],
            }
          }
          title="Sort By Login"
        />
      </div>
    </div>
  </div>
</div>
`;
